Update metadata sync guidelines (#12004)
Alluxio/alluxio: docs/en/core-services/Unified-Namespace.md

data for that file, and reload the metadata for the updated file.
If a file is added or deleted in the UFS, Alluxio will update the metadata in its namespace as well.

The fingerprint is verified with the UFS based on the
`alluxio.user.file.metadata.sync.interval` client configuration property.

For example, if a client executes an operation with the interval set to one minute,
the relevant metadata will be refreshed from the UFS if the last refresh was over a minute ago.
A value of `0` indicates that the metadata will always be synced for every operation,
whereas the default value of `-1` indicates the metadata is never synced again after the initial load.

The graph below illustrates the impact of setting the metadata sync interval on the latency of
client operations.

<p align="center">
<img style="text-align: center" src="{{ '/img/unified-namespace-metadata-sync-performance.png' | relativize_url }}" alt="Metadata sync interval"/>
</p>

The overall cost of syncing with sync interval enabled (blue line) is amortized across the number of
RPCs within the sync time interval. When sync interval enabled to some non-0 value users should

> expect performance spikes in normal application usage. The frequency at which the spikes occur will

Reviewer: ```suggestion
expect latency spikes in normal application usage. The frequency at which the spikes occur will
```

Author: Updated, thanks.

#### All operations through Alluxio except for a select few

##### HDFS UFS

With HDFS, if most operations except for some special cases are updated through Alluxio, we can
take two approaches. One is [active sync](#active-sync-for-hdfs). The other is to set the sync
interval using path configuration.

If HDFS is updated very frequently (not on the Alluxio mount point) and there is a large namespace,
it is more desirable to reduce load on the namenode by disabling active sync and simply using path
configuration instead. Otherwise, if the load on the namenode isn't a concern, use ActiveSync and
set the sync interval to -1.

##### Non-HDFS UFS

If a small number of paths are updated regularly or semi-regularly through the UFS directly instead
of Alluxio, we recommend using the path configuration feature to set the sync interval on the
specific path(s) based on the understanding of the rate of change in the UFS. This allows files to
be synced for the special cases without detrimentally affecting the performance of operations on
other parts of the namespace.

#### UFS frequently updated not through Alluxio

##### HDFS

> If the updates happen across many parts (or unknown) of the namespace at a frequent rate, use active

Reviewer: ```suggestion
If the updates happen across many (or unknown) parts of the namespace at a frequent rate, use active
```

Author: Updated, thanks.

sync in order to update metadata in a timely manner. If the rate of change is too fast, namespace is
too large, or the user does not want to update the max RPC size from the namenode, then follow the
suggestion in the section below for Non-HDFS UFS.

##### Non-HDFS UFS

Set the sync interval for each subtree or the entire cluster based on the understanding of the rate
of change in the UFS.

### Other Methods for Loading New UFS Files

We recommend to configure the sync interval exclusively. It is not recommended to consider
configuring the metadata load type. However, do note the following equivalencies

`alluxio.user.file.metadata.load.type=NEVER` is equivalent to `alluxio.user.metadata.sync.interval=-1`
`alluxio.user.file.metadata.load.type=ALWAYS` is equivalent to `alluxio.user.metadata.sync.interval=0`

If the metadata sync interval is configured the metadata load type is ignored.

### Active Sync for HDFS

> In version 2.0, we introduced a new feature for maintaining synchronization between Alluxio space

Reviewer: nit: newline above this line to be consistent?

Author: Updated, thanks.

To disable active sync on a directory, issue the following Alluxio command.

```console
$ ./bin/alluxio fs stopSync /syncdir
```

> Note: When `startSync` is issued, a full scan of the sync point is scheduled.
> If run as the Alluxio superuser, `stopSync` will interrupt any full scans which have not yet ended.
> If run as any other user, `stopSync` will wait for the full scan to finish before completing.

You can also examine which directories are currently under active sync, with the following command.

```console
$ ./bin/alluxio fs getSyncPathList
```

#### Quiet period for Active Sync

Active sync also tries to avoid syncing when the target directory is heavily used.
It tries to look for a quiet period in UFS activity to start syncing between the UFS and the Alluxio space, to avoid overloading the UFS when it is busy.
There are two configuration options that control this behavior.

`alluxio.master.ufs.active.sync.max.activities` is the maximum number of activities in the UFS directory.
Activity is a heuristic based on the exponential moving average of number of events in a directory.

> For example, if a directory had 100, 10, 1 event in the past three intervals.

Reviewer: ```suggestion
For example, if a directory had 100, 10, and 1 events in the past three intervals.
```

Author: Updated, thanks
